File: 2-Module/2-week/5-day/class-employee-manager-long-practice/manager.js
const employeeClass = require('./employee');
console.log(employeeClass)
class Manager extends employeeClass {
    constructor (name,salary,title,manager = null){
        super(name,salary,title,manager )
        // i feel this part is not very dry, but if I delete manager constructor name,
        // the super won't work
            this.employees = []
        }


        addEmployee(employee){
             this.employees.push(employee);

        }

        // the employees array is an array of instances;
        // add all subsalary together
        // we loop thru the employee array. if employee object is not manager, then we simple
        // add their salary to the sum. Otherwise, we call totalSubsalary method again.




        _totalSubSalary=()=>{
            console.log('hello0')
            // base case;
            let sum = 0;
            // using for loop
            console.log(this.employees.length)
            for(let i=0; i< this.employees.length; i++){
                console.log(i)
              let currentInstance = this.employees[i];
              if(currentInstance instanceof Manager){
                  console.log('hello1')
                  sum += currentInstance.salary + currentInstance._totalSubSalary()
              }
              else{
                  console.log('hello2')
                  sum += currentInstance.salary
              }
            }
            return sum
        }

        calculateBonus(multiplier){
            return (this.salary + this._totalSubSalary())*multiplier
        }

    }

module.exports = Manager;

File: 6-Module/2-week/4-day/employee_manager/manager.py
from employee import Employee

class Manager(Employee):
    def __init__(self, name, salary, title, manager):
        super().__init__(name, salary, title, manager)
        self.employees = []

    def add_employee(self, employee):
        self.employees.append(employee)
        return employee

    def calculate_bonus(self, multiplier):
        return (self.salary + self.total_sub_salary()) * multiplier

    def total_sub_salary(self):
        total_sub_salary = 0

        for employee in self.employees:
            if isinstance(employee, Manager):
                total_sub_salary += employee.salary + employee.total_sub_salary()
            else:
                total_sub_salary += employee.salary

        return total_sub_salary
